### PR TITLE
Add pass to convert logical objectfifo produce/consume to acquire/release

### PR DESCRIPTION
**Pull request by jtuyls**
Adds a pass to convert logical objectfifo consume/produce operations to acquire/release semaphore operations. These acquire/release operations can then later be converted into objectfifo acquire/release operations in the AIE dialect.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  return getParentOpInBlock(block, parentOp);
+}
+
+/// Walk all consume/produce operations within the core operations and insert
+/// semaphore operations.
+template <typename OpTy>
+LogicalResult consumeProduceToAcquireRelease(Operation *parentOp) {
+  using IteratorType = std::conditional_t<
+      std::is_same<OpTy, AMDAIE::LogicalObjectFifoConsume>::value,
+      ForwardIterator, ReverseIterator>;
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp:32]: Why do you need both ForwardIterator and ReverseIterator?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      // it was encountered before. Add it to the map and advance if not.
+      Operation *dmaOp = op.getDma().getDefiningOp();
+      if (!mapper.contains(dmaOp)) {
+        mapper.map(dmaOp, op.getOperation());
+        return WalkResult::advance();
+      }
+
+      // Find the new consume/produce operation's parent operation within the
+      // same block as the previous operation of the same type and operating on
+      // the same DMA. Use theisparent operation in the same block to set the
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp:58]: Typo: Use this parent

**Review by yzhang93 — APPROVED** [review]
Overall LGTM. Thanks!

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp:32]: For the produce operation, a `release` operation needs to be inserted at the place where the produce was as the data will be 'produced' and then the data object is released. And an acquire operation needs to be inserted earlier in the block to acquire the data object resource so we can produce data into it. To accomplish this, the `ReverseIterator` is more convenient. For the consume operator it's the other way around as an acquire operation needs to be inserted at the location of the consume operation so the data object can be read and the release needs to be inserted after the data is not needed anymore.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp (hunk 2 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConsumeProduceToAcquireRelease.cpp]: Done

**Pull request merged**